spectrum-offchain/src/box_resolver/process.rs

```diff
@@ -0,0 +1,75 @@
+use std::pin::Pin;
+
+use futures::channel::mpsc::UnboundedReceiver;
+use futures::stream::select_all;
+use futures::{Stream, StreamExt};
+
+use crate::box_resolver::persistence::EntityRepo;
+use crate::data::reprod_entity::events::{Upgrade, UpgradeRollback};
+use crate::data::reprod_entity::state::{Confirmed, Unconfirmed};
+use crate::data::Has;
+
+pub fn box_tracker_stream<'a, TRepo, TEntity, TEntityId, TStateId>(
+    persistence: TRepo,
+    conf_upgrades: UnboundedReceiver<Upgrade<Confirmed<TEntity>>>,
+    unconf_upgrades: UnboundedReceiver<Upgrade<Unconfirmed<TEntity>>>,
+    rollbacks: UnboundedReceiver<UpgradeRollback<TEntity>>,
+) -> impl Stream<Item = ()> + 'a
+where
+    TRepo: EntityRepo<TEntity, TEntityId, TStateId> + Clone + 'a,
+    TEntity: Has<TEntityId> + Has<TStateId> + 'a,
+{
+    select_all(vec![
+        track_conf_upgrades(persistence.clone(), conf_upgrades),
+        track_unconf_upgrades(persistence.clone(), unconf_upgrades),
+        handle_rollbacks(persistence.clone(), rollbacks),
```

Comment:
Redundant clone

Reply:
This piece is rewritten in the subsequent PR